pkg/openapiconv/convert.go

```diff
@@ -0,0 +1,313 @@
+var OpenAPIV3DefPrefix = "#/components/schemas/"
+
+// ConvertV2ToV3 converts an OpenAPI V2 object into V3.
+// Certain references may be shared between the V2 and V3 objects in the conversion.
+func ConvertV2ToV3(v2Spec *spec.Swagger) (*spec3.OpenAPI) {
+	v3Spec := &spec3.OpenAPI{
+		Version:      "3.0.0",
+		Info:         v2Spec.Info,
+		ExternalDocs: ConvertExternalDocumentation(v2Spec.ExternalDocs),
+		Paths:        ConvertPaths(v2Spec.Paths),
+		Components:   ConvertComponents(v2Spec.SecurityDefinitions, v2Spec.Definitions, v2Spec.Responses, v2Spec.Produces),
+	}
+
+	return v3Spec
+}
+
+func ConvertExternalDocumentation(v2ED *spec.ExternalDocumentation) *spec3.ExternalDocumentation {
+	if v2ED == nil {
+		return nil
+	}
+	return &spec3.ExternalDocumentation{
+		ExternalDocumentationProps: spec3.ExternalDocumentationProps{
+			Description: v2ED.Description,
+			URL:         v2ED.URL,
+		},
+	}
+}
+
+func ConvertComponents(v2SecurityDefinitions spec.SecurityDefinitions, v2Definitions spec.Definitions, v2Responses map[string]spec.Response, produces []string) *spec3.Components {
+	components := &spec3.Components{}
+
+	if v2Definitions != nil {
+		components.Schemas = make(map[string]*spec.Schema)
+	}
+	for s, schema := range v2Definitions {
+		components.Schemas[s] = ConvertSchema(&schema)
+	}
+	if v2SecurityDefinitions != nil {
+		components.SecuritySchemes = make(spec3.SecuritySchemes)
+	}
+	for s, securityScheme := range v2SecurityDefinitions {
+		components.SecuritySchemes[s] = ConvertSecurityScheme(securityScheme)
+	}
+	if v2Responses != nil {
+		components.Responses = make(map[string]*spec3.Response)
+	}
+	for r, response := range v2Responses {
+		components.Responses[r] = ConvertResponse(&response, produces)
+	}
+
+	return components
+}
+
+func ConvertSchema(v2Schema *spec.Schema) *spec.Schema {
+	if v2Schema == nil {
+		return nil
+	}
+	v3Schema := spec.Schema{
+		VendorExtensible: v2Schema.VendorExtensible,
+		SchemaProps: v2Schema.SchemaProps,
+		SwaggerSchemaProps: v2Schema.SwaggerSchemaProps,
+		ExtraProps: v2Schema.ExtraProps,
+	}
+
+	if refString := v2Schema.Ref.String(); refString != "" {
+		if idx := strings.Index(refString, OpenAPIV2DefPrefix); idx != -1 {
+			v3Schema.Ref = spec.MustCreateRef(OpenAPIV3DefPrefix + refString[idx+len(OpenAPIV2DefPrefix):])
+		} else {
+			klog.Errorf("Error: Swagger V2 Ref %s does not contain #/definitions\n", refString)
+		}
+	}
+
+	if v2Schema.Properties != nil {
+		v3Schema.Properties = make(map[string]spec.Schema)
+		for key, property := range v2Schema.Properties {
+			v3Schema.Properties[key] = *ConvertSchema(&property)
+		}
+	}
+	if v2Schema.Items != nil {
+		v3Schema.Items = new(spec.SchemaOrArray)
+		v3Schema.Items.Schema = ConvertSchema(v2Schema.Items.Schema)
+		if v2Schema.Items.Schemas != nil {
+			v3Schema.Items.Schemas = []spec.Schema{}
+			for _, s := range v2Schema.Items.Schemas {
+				v3Schema.Items.Schemas = append(v3Schema.Items.Schemas, *ConvertSchema(&s))
+			}
+		}
```

Comment:
Not something I specifically want you to change, but you could have written this like:
```
v3Schema.Items := spec.SchemaOrArray {
    Schema: ConvertSchema(v2schema.Items.Schema),
    Schemas: ConvertSchemaList(v2.schema.Items.Schemas), // Move the for loop into its own function
}
```

Reply:
You could have just said you want me to change it :)

Updated.